// koa-simple-auth/index.js
"use strict";
var mongoose = require('mongoose');

module.exports = function *authenticate(next){
       
  var User = module.exports;
  try{
    if(!this.session) throw new Error('no session');
    if(!this.session.uid) throw new Error('user id not set');
    this.state.user = yield User.model.findById(this.session.uid)
    .select('-passwordhash -__v')
    .exec()
    .then(
      function(user){
        if(!user) throw new Error('user not found');
        return user;
      },
      function(err){ throw err }
    );
      yield next;
  } catch(err){
    delete this.session.uid;
    delete this.state.user;
    this.status = 401;
    //this.app.emit('error', err, this);
  }
};

var db_url = "mongodb://localhost/users";
module.exports.db = mongoose.createConnection(db_url);
module.exports.login = require('./lib/login');
module.exports.logout = require('./lib/logout');
module.exports.model = require('./lib/model');
module.exports.register = require('./lib/register');
module.exports.unregister = require('./lib/unregister');
